src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Work(Uint256);
+
+impl Work {
+    /// Lowest possible work value for Mainnet. See comment on [`Params::pow_limit`] for more info.
+    pub const MAINNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0000_ffff_0000_u64]));
+
+    /// Lowest possible work value for Testnet. See comment on [`Params::pow_limit`] for more info.
+    pub const TESTNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0000_ffff_0000_u64]));
+
+    /// Lowest possible work value for Signet. See comment on [`Params::pow_limit`] for more info.
+    pub const SIGNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0377_ae00_0000_u64]));
+
+    /// Lowest possible work value for Regtest. See comment on [`Params::pow_limit`] for more info.
+    pub const REGTEST_MIN: Work = Work(Uint256([0, 0, 0, 0x7fff_ff00_0000_0000_u64]));
+
+    /// Converts this [`Work`] to [`Target`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(&self) -> Target { Target(self.0.inverse()) }
+}
+
+impl Add for Work {
+    type Output = Work;
+    fn add(self, rhs: Self) -> Self { Work(self.0.wrapping_add(rhs.0)) }
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
+
+do_impl!(Work);
+
+impl_mul!(Work, u8);
+impl_mul!(Work, u16);
+impl_mul!(Work, u32);
+
+impl_div!(Work, u8);
+impl_div!(Work, u16);
+impl_div!(Work, u32);
+
+/// A 256 bit integer representing target.
+///
+/// The SHA-256 hash of a block's header must be lower than or equal to the current target for the
+/// block to be accepted by the network. The lower the target, the more difficult it is to generate
+/// a block. (See also [`Work`].)
+///
+/// ref: <https://en.bitcoin.it/wiki/Target>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Target;
+/// let _ = Target::from(0xdeadbeef);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Target(Uint256);
+
+impl Target {
+    /// Computes the popular "difficulty" measure for mining.
+    ///
+    /// Difficulty represents how difficult the current target makes it to find a block, relative to
+    /// how difficult it would be at the highest possible target (highest target == lowest difficulty).
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    pub fn difficulty(&self) -> u64 {
```

Comment:
Should contain documentation about too low target causing overflow. But wouldn't it be better to return `None` if difficulty is too high? We could also return `u128` but that still overflows. Or perhaps saturating is better than potentially returning too low number.

Reply:
I played around with this but have not come up with a good solution yet.

Reply:
I've implemented saturation and added docs, not sure if this is the best solution.